{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import time\n",
    "from gseapy.plot import barplot, dotplot\n",
    "\n",
    "wk_dir = '/Users/jplab/Desktop/DAILY_CODE_DATA/2022-3/data/3-15_DMSO_TGF_BvsTshnc/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### need to set species here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ‘Human’, ‘Mouse’, ‘Yeast’, ‘Fly’, ‘Fish’, ‘Worm’\n",
    "speices = 'Human'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* add file path\n",
    "* add df.name\n",
    "* append to list\n",
    "* **NOTE** need to pay attention to gene name in tsv and up down case issue\n",
    "* Seems enricr has a list **size limit** and will casuse issue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df_DMSO_nc_vs_kd\n",
      "df_TGF_nc_vs_kd\n",
      "df_shFOXA1_B14_vs_TGF\n",
      "df_shNC_B14_vs_TGF\n",
      "dict_keys(['df_DMSO_nc_vs_kd_UP_genes', 'df_DMSO_nc_vs_kd_DOWN_genes', 'df_TGF_nc_vs_kd_UP_genes', 'df_TGF_nc_vs_kd_DOWN_genes', 'df_shFOXA1_B14_vs_TGF_UP_genes', 'df_shFOXA1_B14_vs_TGF_DOWN_genes', 'df_shNC_B14_vs_TGF_UP_genes', 'df_shNC_B14_vs_TGF_DOWN_genes'])\n",
      "df_DMSO_nc_vs_kd_UP_genes: 0 rows of dup are dropped\n",
      "df_DMSO_nc_vs_kd_UP_genes: 0 rows of NA are dropped\n",
      "df_DMSO_nc_vs_kd_DOWN_genes: 0 rows of dup are dropped\n",
      "df_DMSO_nc_vs_kd_DOWN_genes: 0 rows of NA are dropped\n",
      "df_TGF_nc_vs_kd_UP_genes: 0 rows of dup are dropped\n",
      "df_TGF_nc_vs_kd_UP_genes: 0 rows of NA are dropped\n",
      "df_TGF_nc_vs_kd_DOWN_genes: 0 rows of dup are dropped\n",
      "df_TGF_nc_vs_kd_DOWN_genes: 0 rows of NA are dropped\n",
      "df_shFOXA1_B14_vs_TGF_UP_genes: 0 rows of dup are dropped\n",
      "df_shFOXA1_B14_vs_TGF_UP_genes: 0 rows of NA are dropped\n",
      "df_shFOXA1_B14_vs_TGF_DOWN_genes: 1 rows of dup are dropped\n",
      "df_shFOXA1_B14_vs_TGF_DOWN_genes: 0 rows of NA are dropped\n",
      "df_shNC_B14_vs_TGF_UP_genes: 0 rows of dup are dropped\n",
      "df_shNC_B14_vs_TGF_UP_genes: 0 rows of NA are dropped\n",
      "df_shNC_B14_vs_TGF_DOWN_genes: 0 rows of dup are dropped\n",
      "df_shNC_B14_vs_TGF_DOWN_genes: 0 rows of NA are dropped\n"
     ]
    }
   ],
   "source": [
    "# add file path\n",
    "# add df.name\n",
    "# append to list\n",
    "DEG_DMSO_nc_vs_kd = '/Users/jplab/Desktop/snakepipes_Data/results_mRNA_seq_FOXA1_KD_NC_2022221/DESeq2_samplesheet_FOXA1-DMSO-NCvsKD/DEseq_basic_DEresults_counts_DESeq2.normalized_noVersion.tsv'\n",
    "DEG_TGF_nc_vs_kd = '/Users/jplab/Desktop/snakepipes_Data/results_mRNA_seq_FOXA1_KD_NC_2022221/DESeq2_samplesheet_FOXA1-TGFb-NCvsKD/DEseq_basic_DEresults_counts_DESeq2.normalized_noVersion.tsv'\n",
    "DEG_shFOXA1_B14_vs_TGF = '/Users/jplab/Desktop/snakepipes_Data/results_mRNA_seq_FOXA1_KD_NC_2022221/DESeq2_samplesheet_FOXA1KD-B14vsTGFb/DEseq_basic_DEresults_counts_DESeq2.normalized_noVersion.tsv'\n",
    "DEG_shNC_B14_vs_TGF = '/Users/jplab/Desktop/snakepipes_Data/results_mRNA_seq_FOXA1_KD_NC_2022221/DESeq2_samplesheet_shNC-B14vsTGFb/DEseq_basic_DEresults_counts_DESeq2.normalized_noVersion.tsv'\n",
    "\n",
    "\n",
    "df_DMSO_nc_vs_kd = pd.read_csv(DEG_DMSO_nc_vs_kd,header=0, sep=\"\\t\",index_col=False)\n",
    "df_DMSO_nc_vs_kd.name = 'df_DMSO_nc_vs_kd'\n",
    "\n",
    "df_TGF_nc_vs_kd = pd.read_csv(DEG_TGF_nc_vs_kd,header=0, sep=\"\\t\",index_col=False)\n",
    "df_TGF_nc_vs_kd.name = 'df_TGF_nc_vs_kd'\n",
    "\n",
    "df_shFOXA1_B14_vs_TGF = pd.read_csv(DEG_shFOXA1_B14_vs_TGF,header=0, sep=\"\\t\",index_col=False)\n",
    "df_shFOXA1_B14_vs_TGF.name = 'df_shFOXA1_B14_vs_TGF'\n",
    "\n",
    "df_shNC_B14_vs_TGF = pd.read_csv(DEG_shNC_B14_vs_TGF,header=0, sep=\"\\t\",index_col=False)\n",
    "df_shNC_B14_vs_TGF.name = 'df_shNC_B14_vs_TGF'\n",
    "\n",
    "df_list = []\n",
    "df_list.append(df_DMSO_nc_vs_kd)\n",
    "df_list.append(df_TGF_nc_vs_kd)\n",
    "df_list.append(df_shFOXA1_B14_vs_TGF)\n",
    "df_list.append(df_shNC_B14_vs_TGF)\n",
    "\n",
    "# first use dict to save loop filter results\n",
    "dict = {}\n",
    "for df in df_list:\n",
    "    print(df.name)\n",
    "    df_name = df.name\n",
    "\n",
    "    dict[df_name + '_UP_genes'] = df.query(\"Status == 'UP'\")\n",
    "    dict[df_name + '_DOWN_genes'] = df.query(\"Status == 'DOWN'\")\n",
    "    # dict[df_name + '_UP_and_DOWN_genes'] = df.query(\"change == 'Down' or change == 'Up'\")\n",
    "print(dict.keys())\n",
    "\n",
    "# then loop dict\n",
    "dict_gene_list = {}\n",
    "for df_name, df in dict.items():\n",
    "    # sort by fold change\n",
    "    df = df.sort_values(by=['log2FoldChange'], ascending=False)\n",
    "\n",
    "    # drop dup in symbol column\n",
    "    df_ori_row_num = len(df.index)\n",
    "    df = df.drop_duplicates(subset=['external_gene_name'], keep='first')\n",
    "    df_dupDrop_row_num = len(df.index)\n",
    "    print(df_name + ': ' + str(df_ori_row_num - df_dupDrop_row_num) + ' rows of dup are dropped')\n",
    "\n",
    "    # must drop na, or will throw error in enrichr method\n",
    "    df = df['external_gene_name'].dropna()\n",
    "    df_dropNa_row_num = len(df.index)\n",
    "    print(df_name + ': ' + str(df_dupDrop_row_num - df_dropNa_row_num) + ' rows of NA are dropped')\n",
    "\n",
    "    # convert to list\n",
    "    dict_gene_list[df_name.split('df_')[1]] = df.tolist()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ARCHS4_Cell-lines', 'ARCHS4_IDG_Coexp', 'ARCHS4_Kinases_Coexp', 'ARCHS4_TFs_Coexp', 'ARCHS4_Tissues', 'Achilles_fitness_decrease', 'Achilles_fitness_increase', 'Aging_Perturbations_from_GEO_down', 'Aging_Perturbations_from_GEO_up', 'Allen_Brain_Atlas_10x_scRNA_2021', 'Allen_Brain_Atlas_down', 'Allen_Brain_Atlas_up', 'Azimuth_Cell_Types_2021', 'BioCarta_2013', 'BioCarta_2015', 'BioCarta_2016', 'BioPlanet_2019', 'BioPlex_2017', 'CCLE_Proteomics_2020', 'CORUM', 'COVID-19_Related_Gene_Sets', 'COVID-19_Related_Gene_Sets_2021', 'Cancer_Cell_Line_Encyclopedia', 'CellMarker_Augmented_2021', 'ChEA_2013', 'ChEA_2015', 'ChEA_2016', 'Chromosome_Location', 'Chromosome_Location_hg19', 'ClinVar_2019', 'DSigDB', 'Data_Acquisition_Method_Most_Popular_Genes', 'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019', 'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019', 'Descartes_Cell_Types_and_Tissue_2021', 'DisGeNET', 'Disease_Perturbations_from_GEO_down', 'Disease_Perturbations_from_GEO_up', 'Disease_Signatures_from_GEO_down_2014', 'Disease_Signatures_from_GEO_up_2014', 'DrugMatrix', 'Drug_Perturbations_from_GEO_2014', 'Drug_Perturbations_from_GEO_down', 'Drug_Perturbations_from_GEO_up', 'ENCODE_Histone_Modifications_2013', 'ENCODE_Histone_Modifications_2015', 'ENCODE_TF_ChIP-seq_2014', 'ENCODE_TF_ChIP-seq_2015', 'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X', 'ESCAPE', 'Elsevier_Pathway_Collection', 'Enrichr_Libraries_Most_Popular_Genes', 'Enrichr_Submissions_TF-Gene_Coocurrence', 'Enrichr_Users_Contributed_Lists_2020', 'Epigenomics_Roadmap_HM_ChIP-seq', 'GO_Biological_Process_2013', 'GO_Biological_Process_2015', 'GO_Biological_Process_2017', 'GO_Biological_Process_2017b', 'GO_Biological_Process_2018', 'GO_Biological_Process_2021', 'GO_Cellular_Component_2013', 'GO_Cellular_Component_2015', 'GO_Cellular_Component_2017', 'GO_Cellular_Component_2017b', 'GO_Cellular_Component_2018', 'GO_Cellular_Component_2021', 'GO_Molecular_Function_2013', 'GO_Molecular_Function_2015', 'GO_Molecular_Function_2017', 'GO_Molecular_Function_2017b', 'GO_Molecular_Function_2018', 'GO_Molecular_Function_2021', 'GTEx_Aging_Signatures_2021', 'GTEx_Tissue_Expression_Down', 'GTEx_Tissue_Expression_Up', 'GWAS_Catalog_2019', 'GeneSigDB', 'Gene_Perturbations_from_GEO_down', 'Gene_Perturbations_from_GEO_up', 'Genes_Associated_with_NIH_Grants', 'Genome_Browser_PWMs', 'HDSigDB_Human_2021', 'HDSigDB_Mouse_2021', 'HMDB_Metabolites', 'HMS_LINCS_KinomeScan', 'HomoloGene', 'HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression', 'HumanCyc_2015', 'HumanCyc_2016', 'Human_Gene_Atlas', 'Human_Phenotype_Ontology', 'InterPro_Domains_2019', 'Jensen_COMPARTMENTS', 'Jensen_DISEASES', 'Jensen_TISSUES', 'KEA_2013', 'KEA_2015', 'KEGG_2013', 'KEGG_2015', 'KEGG_2016', 'KEGG_2019_Human', 'KEGG_2019_Mouse', 'KEGG_2021_Human', 'Kinase_Perturbations_from_GEO_down', 'Kinase_Perturbations_from_GEO_up', 'L1000_Kinase_and_GPCR_Perturbations_down', 'L1000_Kinase_and_GPCR_Perturbations_up', 'LINCS_L1000_Chem_Pert_down', 'LINCS_L1000_Chem_Pert_up', 'LINCS_L1000_Ligand_Perturbations_down', 'LINCS_L1000_Ligand_Perturbations_up', 'Ligand_Perturbations_from_GEO_down', 'Ligand_Perturbations_from_GEO_up', 'MCF7_Perturbations_from_GEO_down', 'MCF7_Perturbations_from_GEO_up', 'MGI_Mammalian_Phenotype_2013', 'MGI_Mammalian_Phenotype_2017', 'MGI_Mammalian_Phenotype_Level_3', 'MGI_Mammalian_Phenotype_Level_4', 'MGI_Mammalian_Phenotype_Level_4_2019', 'MGI_Mammalian_Phenotype_Level_4_2021', 'MSigDB_Computational', 'MSigDB_Hallmark_2020', 'MSigDB_Oncogenic_Signatures', 'Microbe_Perturbations_from_GEO_down', 'Microbe_Perturbations_from_GEO_up', 'Mouse_Gene_Atlas', 'NCI-60_Cancer_Cell_Lines', 'NCI-Nature_2015', 'NCI-Nature_2016', 'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions', 'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions', 'NIH_Funded_PIs_2017_Human_AutoRIF', 'NIH_Funded_PIs_2017_Human_GeneRIF', 'NURSA_Human_Endogenous_Complexome', 'OMIM_Disease', 'OMIM_Expanded', 'Old_CMAP_down', 'Old_CMAP_up', 'Orphanet_Augmented_2021', 'PPI_Hub_Proteins', 'PanglaoDB_Augmented_2021', 'Panther_2015', 'Panther_2016', 'Pfam_Domains_2019', 'Pfam_InterPro_Domains', 'PheWeb_2019', 'PhenGenI_Association_2021', 'Phosphatase_Substrates_from_DEPOD', 'ProteomicsDB_2020', 'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO', 'RNAseq_Automatic_GEO_Signatures_Human_Down', 'RNAseq_Automatic_GEO_Signatures_Human_Up', 'RNAseq_Automatic_GEO_Signatures_Mouse_Down', 'RNAseq_Automatic_GEO_Signatures_Mouse_Up', 'Rare_Diseases_AutoRIF_ARCHS4_Predictions', 'Rare_Diseases_AutoRIF_Gene_Lists', 'Rare_Diseases_GeneRIF_ARCHS4_Predictions', 'Rare_Diseases_GeneRIF_Gene_Lists', 'Reactome_2013', 'Reactome_2015', 'Reactome_2016', 'SILAC_Phosphoproteomics', 'SubCell_BarCode', 'SysMyo_Muscle_Gene_Sets', 'TF-LOF_Expression_from_GEO', 'TF_Perturbations_Followed_by_Expression', 'TG_GATES_2020', 'TRANSFAC_and_JASPAR_PWMs', 'TRRUST_Transcription_Factors_2019', 'Table_Mining_of_CRISPR_Studies', 'TargetScan_microRNA', 'TargetScan_microRNA_2017', 'Tissue_Protein_Expression_from_Human_Proteome_Map', 'Tissue_Protein_Expression_from_ProteomicsDB', 'Transcription_Factor_PPIs', 'UK_Biobank_GWAS_v1', 'Virus-Host_PPI_P-HIPSTer_2020', 'VirusMINT', 'Virus_Perturbations_from_GEO_down', 'Virus_Perturbations_from_GEO_up', 'WikiPathway_2021_Human', 'WikiPathways_2013', 'WikiPathways_2015', 'WikiPathways_2016', 'WikiPathways_2019_Human', 'WikiPathways_2019_Mouse', 'dbGaP', 'huMAP', 'lncHUB_lncRNA_Co-Expression', 'miRTarBase_2017']\n"
     ]
    }
   ],
   "source": [
    "# get enrichr lib name from GSEApy\n",
    "lib_names = gp.get_library_name()\n",
    "lib_names\n",
    "print(lib_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set target libs to be used in GSEApy\n",
    "target_libs = [\n",
    "    'WikiPathway_2021_Human',\n",
    "    'KEGG_2021_Human',\n",
    "    'MSigDB_Hallmark_2020',\n",
    "\n",
    "    'GO_Biological_Process_2021',\n",
    "    'GO_Cellular_Component_2021',\n",
    "    'GO_Molecular_Function_2021',\n",
    "    \n",
    "    'GTEx_Tissue_Expression_Down',\n",
    "    'GTEx_Tissue_Expression_Up',\n",
    "    'Kinase_Perturbations_from_GEO_down',\n",
    "    'Kinase_Perturbations_from_GEO_up',\n",
    "    'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "    'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "    'CCLE_Proteomics_2020',\n",
    "    'MSigDB_Oncogenic_Signatures',\n",
    "    'Old_CMAP_down',\n",
    "    'Old_CMAP_up',\n",
    "    'Phosphatase_Substrates_from_DEPOD',\n",
    "    'PPI_Hub_Proteins',\n",
    "    'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "    'RNAseq_Automatic_GEO_Signatures_Human_Down',\n",
    "    'RNAseq_Automatic_GEO_Signatures_Human_Up',\n",
    "    'TRANSFAC_and_JASPAR_PWMs',\n",
    "    'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
    "    ]\n",
    "\n",
    "target_libs = [\n",
    "    'MSigDB_Hallmark_2020',\n",
    "\n",
    "    'GO_Biological_Process_2021',\n",
    "    'GO_Cellular_Component_2021',\n",
    "    'GO_Molecular_Function_2021',\n",
    "\n",
    "    'Mouse_Gene_Atlas',\n",
    "    'HDSigDB_Mouse_2021',\n",
    "    'WikiPathways_2019_Mouse',\n",
    "    'KEGG_2019_Mouse',\n",
    "    'RNAseq_Automatic_GEO_Signatures_Mouse_Down',\n",
    "    'RNAseq_Automatic_GEO_Signatures_Mouse_Up',\n",
    "\n",
    "    'Kinase_Perturbations_from_GEO_down',\n",
    "    'Kinase_Perturbations_from_GEO_up',\n",
    "    'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "    'CCLE_Proteomics_2020',\n",
    "    'MSigDB_Oncogenic_Signatures',\n",
    "    'Old_CMAP_down',\n",
    "    'Old_CMAP_up',\n",
    "    'Phosphatase_Substrates_from_DEPOD',\n",
    "    'PPI_Hub_Proteins',\n",
    "    'TRANSFAC_and_JASPAR_PWMs'\n",
    "    ]      \n",
    "\n",
    "# say something\n",
    "# careful here, if i use test as description, it works; if i use 'nc_B14_vs_TGF and kd_B14_vs_TGF' as description, it will throw error on gp.enrichr method\n",
    "# description = 'nc_B14_vs_TGF and kd_B14_vs_TGF'\n",
    "description = 'test'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WikiPathway_2021_Human\n",
      "DMSO_nc_vs_kd_UP_genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-15 13:46:12,817 Connecting to Enrichr Server to get latest library names\n",
      "2022-03-15 13:46:12,818 Analysis name: test, Enrichr Library: WikiPathway_2021_Human\n",
      "2022-03-15 13:46:19,796 Save file of enrichment results: Job Id:4ce0c892454764280a232488ac561e49\n",
      "2022-03-15 13:46:20,510 Done.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WikiPathway_2021_Human DMSO_nc_vs_kd_UP_genes\n",
      "i got res2d:                    Gene_set  \\\n",
      "0    WikiPathway_2021_Human   \n",
      "1    WikiPathway_2021_Human   \n",
      "2    WikiPathway_2021_Human   \n",
      "3    WikiPathway_2021_Human   \n",
      "4    WikiPathway_2021_Human   \n",
      "..                      ...   \n",
      "501  WikiPathway_2021_Human   \n",
      "502  WikiPathway_2021_Human   \n",
      "503  WikiPathway_2021_Human   \n",
      "504  WikiPathway_2021_Human   \n",
      "505  WikiPathway_2021_Human   \n",
      "\n",
      "                                                  Term Overlap       P-value  \\\n",
      "0    Cholesterol metabolism (includes both Bloch an...   22/46  1.438988e-15   \n",
      "1                Nuclear Receptors Meta-Pathway WP2882  52/319  2.074331e-11   \n",
      "2               Cholesterol Biosynthesis Pathway WP197   11/15  3.073607e-11   \n",
      "3                                  NRF2 pathway WP2884  31/146  3.420795e-10   \n",
      "4              p53 transcriptional gene network WP4963   19/67  5.901642e-09   \n",
      "..                                                 ...     ...           ...   \n",
      "501      22q11.2 copy number variation syndrome WP4657   3/131  9.851747e-01   \n",
      "502                            Joubert Syndrome WP4656    1/76  9.901306e-01   \n",
      "503                                Ciliopathies WP4803   4/183  9.952456e-01   \n",
      "504                              mRNA Processing WP411   2/126  9.958169e-01   \n",
      "505                GPCRs, Class A Rhodopsin-like WP455   1/257  9.999960e-01   \n",
      "\n",
      "     Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
      "0        7.281280e-13            0                     0   14.919841   \n",
      "1        5.184151e-09            0                     0    3.212358   \n",
      "2        5.184151e-09            0                     0   44.384434   \n",
      "3        4.327306e-08            0                     0    4.400546   \n",
      "4        5.972461e-07            0                     0    6.417764   \n",
      "..                ...          ...                   ...         ...   \n",
      "501      9.930247e-01            0                     0    0.373223   \n",
      "502      9.960359e-01            0                     0    0.212562   \n",
      "503      9.977889e-01            0                     0    0.355180   \n",
      "504      9.977889e-01            0                     0    0.256678   \n",
      "505      9.999960e-01            0                     0    0.061673   \n",
      "\n",
      "     Combined Score                                              Genes  \n",
      "0      5.098831e+02  ABCA1;IDI1;FDPS;SREBF1;HMGCS1;ELOVL5;MSMO1;DHC...  \n",
      "1      7.902014e+01  SLC2A1;SLC2A3;SLC7A11;SLC2A6;CCND1;ENC1;PDK4;A...  \n",
      "2      1.074351e+03  FDPS;IDI1;SQLE;HMGCS1;SC5D;MSMO1;MVD;HMGCR;DHC...  \n",
      "3      9.591421e+01  MGST3;SLC2A1;MGST2;TGFA;SLC2A3;TXN;SLC7A11;SLC...  \n",
      "4      1.216040e+02  BTG2;CDKN1A;SERPINE1;SLC2A1;LIF;SLC7A11;SAT1;B...  \n",
      "..              ...                                                ...  \n",
      "501    5.574581e-03                              SLC25A1;SREBF1;SREBF2  \n",
      "502    2.108289e-03                                              CETN2  \n",
      "503    1.692679e-03                       GALNT11;FLCN;TRAF3IP1;LZTFL1  \n",
      "504    1.075950e-03                                        PTBP2;PSKH1  \n",
      "505    2.440557e-07                                               OPN3  \n",
      "\n",
      "[506 rows x 10 columns]\n",
      "Legend is shown.localtion  best\n",
      "DMSO_nc_vs_kd_DOWN_genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-15 13:46:22,898 Connecting to Enrichr Server to get latest library names\n",
      "2022-03-15 13:46:22,899 Analysis name: test, Enrichr Library: WikiPathway_2021_Human\n",
      "2022-03-15 13:46:31,318 Save file of enrichment results: Job Id:1c866d6eea4276b38e8f60eabf0e9bcd\n",
      "2022-03-15 13:46:31,462 Done.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WikiPathway_2021_Human DMSO_nc_vs_kd_DOWN_genes\n",
      "i got res2d:                    Gene_set  \\\n",
      "0    WikiPathway_2021_Human   \n",
      "1    WikiPathway_2021_Human   \n",
      "2    WikiPathway_2021_Human   \n",
      "3    WikiPathway_2021_Human   \n",
      "4    WikiPathway_2021_Human   \n",
      "..                      ...   \n",
      "499  WikiPathway_2021_Human   \n",
      "500  WikiPathway_2021_Human   \n",
      "501  WikiPathway_2021_Human   \n",
      "502  WikiPathway_2021_Human   \n",
      "503  WikiPathway_2021_Human   \n",
      "\n",
      "                                                  Term Overlap       P-value  \\\n",
      "0                 Retinoblastoma gene in cancer WP2446   45/87  5.152814e-31   \n",
      "1                                mRNA Processing WP411  44/126  1.063983e-21   \n",
      "2                                DNA Replication WP466   25/42  7.744558e-20   \n",
      "3                                     Cell cycle WP179  38/120  2.838497e-17   \n",
      "4                            DNA Mismatch Repair WP531   17/23  2.662550e-16   \n",
      "..                                                 ...     ...           ...   \n",
      "499                                NRF2 pathway WP2884   4/146  9.848383e-01   \n",
      "500          Toll-like Receptor Signaling Pathway WP75   2/103  9.906620e-01   \n",
      "501                                 Peptide GPCRs WP24    1/74  9.921550e-01   \n",
      "502                GPCRs, Class A Rhodopsin-like WP455   5/257  9.997731e-01   \n",
      "503  Metapathway biotransformation Phase I and II W...   2/183  9.999156e-01   \n",
      "\n",
      "     Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
      "0        2.597018e-28            0                     0   16.402246   \n",
      "1        2.681236e-19            0                     0    8.190172   \n",
      "2        1.301086e-17            0                     0   22.179694   \n",
      "3        3.576507e-15            0                     0    7.038770   \n",
      "4        2.683850e-14            0                     0   42.484121   \n",
      "..                ...          ...                   ...         ...   \n",
      "499      9.927170e-01            0                     0    0.414991   \n",
      "500      9.961078e-01            0                     0    0.291907   \n",
      "501      9.961078e-01            0                     0    0.202079   \n",
      "502      9.999156e-01            0                     0    0.290806   \n",
      "503      9.999156e-01            0                     0    0.162188   \n",
      "\n",
      "     Combined Score                                              Genes  \n",
      "0       1143.902421  TOP2A;DNMT1;PCNA;MCM7;SUV39H1;PRIM1;HMGB2;TTK;...  \n",
      "1        395.521982  SF3B4;SF3B3;DHX9;SRSF1;HNRNPU;HNRNPR;PRPF8;PTB...  \n",
      "2        976.011026  PCNA;MCM7;PRIM1;MCM10;POLD3;CDC45;ORC1;POLD1;P...  \n",
      "3        268.181872  PCNA;MCM7;TTK;SMC3;PKMYT1;ANAPC11;CDC20;CCNB2;...  \n",
      "4       1523.568813  RFC5;RFC3;PCNA;RFC4;RFC1;LIG1;RFC2;RPA1;MLH1;P...  \n",
      "..              ...                                                ...  \n",
      "499        0.006340                              EGR1;TGFB1;SRXN1;FTH1  \n",
      "500        0.002739                                         MAP2K4;FOS  \n",
      "501        0.001592                                             BDKRB2  \n",
      "502        0.000066                      F2R;BDKRB2;F2RL1;ADRA1D;LTB4R  \n",
      "503        0.000014                                       NNMT;CYP24A1  \n",
      "\n",
      "[504 rows x 10 columns]\n",
      "Legend is shown.localtion  best\n",
      "TGF_nc_vs_kd_UP_genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-15 13:46:33,721 Connecting to Enrichr Server to get latest library names\n",
      "2022-03-15 13:46:33,722 Analysis name: test, Enrichr Library: WikiPathway_2021_Human\n",
      "2022-03-15 13:46:39,579 Save file of enrichment results: Job Id:00033381313e6d5766df508f3eb94ea7\n",
      "2022-03-15 13:46:39,716 Done.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WikiPathway_2021_Human TGF_nc_vs_kd_UP_genes\n",
      "i got res2d:                    Gene_set  \\\n",
      "0    WikiPathway_2021_Human   \n",
      "1    WikiPathway_2021_Human   \n",
      "2    WikiPathway_2021_Human   \n",
      "3    WikiPathway_2021_Human   \n",
      "4    WikiPathway_2021_Human   \n",
      "..                      ...   \n",
      "455  WikiPathway_2021_Human   \n",
      "456  WikiPathway_2021_Human   \n",
      "457  WikiPathway_2021_Human   \n",
      "458  WikiPathway_2021_Human   \n",
      "459  WikiPathway_2021_Human   \n",
      "\n",
      "                                                  Term Overlap       P-value  \\\n",
      "0    Cholesterol metabolism (includes both Bloch an...   19/46  1.291241e-14   \n",
      "1               Cholesterol Biosynthesis Pathway WP197   10/15  5.184094e-11   \n",
      "2              p53 transcriptional gene network WP4963   17/67  2.316458e-09   \n",
      "3                Nuclear Receptors Meta-Pathway WP2882  38/319  1.494474e-08   \n",
      "4                                  NRF2 pathway WP2884  22/146  3.041212e-07   \n",
      "..                                                 ...     ...           ...   \n",
      "455      22q11.2 copy number variation syndrome WP4657   2/131  9.787866e-01   \n",
      "456               Cytoplasmic Ribosomal Proteins WP477    1/89  9.803516e-01   \n",
      "457  Electron Transport Chain (OXPHOS system in mit...   1/103  9.894271e-01   \n",
      "458            DNA Repair Pathways Full Network WP4946   1/120  9.950205e-01   \n",
      "459                GPCRs, Class A Rhodopsin-like WP455   3/257  9.990849e-01   \n",
      "\n",
      "     Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
      "0        5.939710e-12            0                     0   15.953122   \n",
      "1        1.192342e-08            0                     0   44.913146   \n",
      "2        3.551903e-07            0                     0    7.680379   \n",
      "3        1.718645e-06            0                     0    3.094729   \n",
      "4        2.797915e-05            0                     0    4.016014   \n",
      "..                ...          ...                   ...         ...   \n",
      "455      9.867872e-01            0                     0    0.342690   \n",
      "456      9.867872e-01            0                     0    0.251425   \n",
      "457      9.937478e-01            0                     0    0.216757   \n",
      "458      9.971883e-01            0                     0    0.185625   \n",
      "459      9.990849e-01            0                     0    0.259650   \n",
      "\n",
      "     Combined Score                                              Genes  \n",
      "0        510.190198  ABCA1;IDI1;FDPS;SREBF1;MVK;HMGCS1;MSMO1;DHCR24...  \n",
      "1       1063.670878  FDPS;IDI1;SQLE;MVK;HMGCS1;MSMO1;HMGCR;DHCR7;LS...  \n",
      "2        152.710708  BTG2;CDKN1A;SERPINE1;SLC2A1;SLC7A11;SAT1;BBC3;...  \n",
      "3         55.763638  ABHD2;SLC2A1;TGFA;SLC7A11;SLC2A6;CYP3A5;PRRG4;...  \n",
      "4         60.263660  ABCC3;GSTM3;ABCC4;SLC2A14;G6PD;TGFB2;GPX2;SLC6...  \n",
      "..              ...                                                ...  \n",
      "455        0.007348                                       ACTA2;SREBF1  \n",
      "456        0.004989                                            RPS6KA6  \n",
      "457        0.002304                                            ATP5IF1  \n",
      "458        0.000927                                              CETN2  \n",
      "459        0.000238                                SUCNR1;HTR1D;ADRA1B  \n",
      "\n",
      "[460 rows x 10 columns]\n",
      "Legend is shown.localtion  best\n",
      "TGF_nc_vs_kd_DOWN_genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-15 13:46:41,991 Connecting to Enrichr Server to get latest library names\n",
      "2022-03-15 13:46:41,992 Analysis name: test, Enrichr Library: WikiPathway_2021_Human\n",
      "2022-03-15 13:46:47,476 Save file of enrichment results: Job Id:02e78c4605ce0499110481957620a974\n",
      "2022-03-15 13:46:47,618 Done.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WikiPathway_2021_Human TGF_nc_vs_kd_DOWN_genes\n",
      "i got res2d:                    Gene_set  \\\n",
      "0    WikiPathway_2021_Human   \n",
      "1    WikiPathway_2021_Human   \n",
      "2    WikiPathway_2021_Human   \n",
      "3    WikiPathway_2021_Human   \n",
      "4    WikiPathway_2021_Human   \n",
      "..                      ...   \n",
      "388  WikiPathway_2021_Human   \n",
      "389  WikiPathway_2021_Human   \n",
      "390  WikiPathway_2021_Human   \n",
      "391  WikiPathway_2021_Human   \n",
      "392  WikiPathway_2021_Human   \n",
      "\n",
      "                                                  Term Overlap       P-value  \\\n",
      "0                 Retinoblastoma gene in cancer WP2446   48/87  4.441613e-45   \n",
      "1                                DNA Replication WP466   25/42  3.448832e-25   \n",
      "2                                     Cell cycle WP179  37/120  9.400460e-24   \n",
      "3    DNA IR-damage and cellular response via ATR WP...   30/80  2.500170e-22   \n",
      "4              DNA Repair Pathways Full Network WP4946  30/120  1.121190e-16   \n",
      "..                                                 ...     ...           ...   \n",
      "388                       Amino Acid metabolism WP3925    1/91  9.714805e-01   \n",
      "389                               Ras signaling WP4223   3/184  9.737559e-01   \n",
      "390            Nonalcoholic fatty liver disease WP4396   2/155  9.833069e-01   \n",
      "391  Angiopoietin Like Protein 8 Regulatory Pathway...   1/132  9.942866e-01   \n",
      "392                GPCRs, Class A Rhodopsin-like WP455   2/257  9.995250e-01   \n",
      "\n",
      "     Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
      "0        1.745554e-42            0                     0   32.950971   \n",
      "1        6.776955e-23            0                     0   38.191574   \n",
      "2        1.231460e-21            0                     0   11.727525   \n",
      "3        2.456417e-20            0                     0   15.661224   \n",
      "4        8.812555e-15            0                     0    8.682540   \n",
      "..                ...          ...                   ...         ...   \n",
      "388      9.812463e-01            0                     0    0.278432   \n",
      "389      9.812463e-01            0                     0    0.414452   \n",
      "390      9.883366e-01            0                     0    0.326917   \n",
      "391      9.968230e-01            0                     0    0.190880   \n",
      "392      9.995250e-01            0                     0    0.195102   \n",
      "\n",
      "     Combined Score                                              Genes  \n",
      "0       3365.128175  TOP2A;DNMT1;PCNA;MCM7;SUV39H1;PRIM1;HMGB2;TTK;...  \n",
      "1       2151.201190  PRIM2;PCNA;MCM7;PRIM1;GMNN;MCM10;POLD3;CDC45;O...  \n",
      "2        621.808455  PCNA;MCM7;TTK;SMC3;PKMYT1;CDC20;CCNB2;CCNB1;CD...  \n",
      "3        778.997346  MDC1;FEN1;PCNA;BRCA1;FOXM1;BRCA2;BRIP1;CDC45;E...  \n",
      "4        318.883380  FEN1;PCNA;HMGB1;BRCA1;BRCA2;POLD3;BRIP1;EXO1;P...  \n",
      "..              ...                                                ...  \n",
      "388        0.008056                                            ALDH7A1  \n",
      "389        0.011022                                     GNG4;NF1;PLCE1  \n",
      "390        0.005503                                      BCL2L11;TGFB1  \n",
      "391        0.001094                                             MAP2K4  \n",
      "392        0.000093                                       LPAR5;ADRA1D  \n",
      "\n",
      "[393 rows x 10 columns]\n",
      "Legend is shown.localtion  best\n",
      "shFOXA1_B14_vs_TGF_UP_genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-15 13:46:49,830 Connecting to Enrichr Server to get latest library names\n",
      "2022-03-15 13:46:49,831 Analysis name: test, Enrichr Library: WikiPathway_2021_Human\n",
      "2022-03-15 13:46:57,426 Save file of enrichment results: Job Id:6534356e7df8b673a267500d021e3f8a\n",
      "2022-03-15 13:46:57,568 Done.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WikiPathway_2021_Human shFOXA1_B14_vs_TGF_UP_genes\n",
      "i got res2d:                    Gene_set  \\\n",
      "0    WikiPathway_2021_Human   \n",
      "1    WikiPathway_2021_Human   \n",
      "2    WikiPathway_2021_Human   \n",
      "3    WikiPathway_2021_Human   \n",
      "4    WikiPathway_2021_Human   \n",
      "..                      ...   \n",
      "541  WikiPathway_2021_Human   \n",
      "542  WikiPathway_2021_Human   \n",
      "543  WikiPathway_2021_Human   \n",
      "544  WikiPathway_2021_Human   \n",
      "545  WikiPathway_2021_Human   \n",
      "\n",
      "                                                  Term Overlap       P-value  \\\n",
      "0    Electron Transport Chain (OXPHOS system in mit...  47/103  9.081564e-24   \n",
      "1                         Amino Acid metabolism WP3925   36/91  4.399285e-16   \n",
      "2              Nonalcoholic fatty liver disease WP4396  47/155  3.000836e-15   \n",
      "3                      Oxidative phosphorylation WP623   25/60  3.394074e-12   \n",
      "4      TCA Cycle (aka Krebs or citric acid cycle) WP78   12/18  1.433740e-09   \n",
      "..                                                 ...     ...           ...   \n",
      "541                              mRNA Processing WP411   5/126  9.843458e-01   \n",
      "542  Arrhythmogenic Right Ventricular Cardiomyopath...    2/74  9.885919e-01   \n",
      "543  Myometrial relaxation and contraction pathways...   6/156  9.925618e-01   \n",
      "544               Cytoplasmic Ribosomal Proteins WP477    2/89  9.964354e-01   \n",
      "545                GPCRs, Class A Rhodopsin-like WP455   2/257  9.999957e-01   \n",
      "\n",
      "     Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
      "0        4.958534e-21            0                     0    9.343195   \n",
      "1        1.201005e-13            0                     0    7.238456   \n",
      "2        5.461521e-13            0                     0    4.830821   \n",
      "3        4.632911e-10            0                     0    7.855424   \n",
      "4        1.565645e-07            0                     0   21.859104   \n",
      "..                ...          ...                   ...         ...   \n",
      "541      9.916103e-01            0                     0    0.446930   \n",
      "542      9.940538e-01            0                     0    0.300709   \n",
      "543      9.962109e-01            0                     0    0.432195   \n",
      "544      9.982638e-01            0                     0    0.248658   \n",
      "545      9.999957e-01            0                     0    0.084054   \n",
      "\n",
      "     Combined Score                                              Genes  \n",
      "0      4.957106e+02  NDUFB9;COX7B;UQCRB;NDUFB10;COX15;NDUFB5;COX4I1...  \n",
      "1      2.559512e+02  FH;HIBADH;PPM1L;EPRS;AUH;PDK4;ACADM;ACAA1;HADH...  \n",
      "2      1.615421e+02  CEBPA;NDUFB9;COX7B;UQCRB;NDUFB10;NDUFB5;COX4I1...  \n",
      "3      2.074538e+02  NDUFB9;NDUFB10;NDUFB5;NDUFB4;NDUFA10;ATP5MC3;A...  \n",
      "4      4.451165e+02  CS;FH;SUCLA2;MDH2;IDH3B;OGDH;SDHD;ACO2;SDHA;SD...  \n",
      "..              ...                                                ...  \n",
      "541    7.051683e-03                 HNRNPA2B1;SNRPA1;SNRPA;PTBP2;SRPK1  \n",
      "542    3.450224e-03                                         PKP2;ITGB8  \n",
      "543    3.226758e-03                CAMK2D;GRK6;LPAR1;ARRB1;ARRB2;GABPA  \n",
      "544    8.879421e-04                                     RPS6KA1;MRPL19  \n",
      "545    3.580694e-07                                         GPR37;OPN3  \n",
      "\n",
      "[546 rows x 10 columns]\n",
      "Legend is shown.localtion  best\n",
      "shFOXA1_B14_vs_TGF_DOWN_genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-15 13:46:59,739 Connecting to Enrichr Server to get latest library names\n",
      "2022-03-15 13:46:59,741 Analysis name: test, Enrichr Library: WikiPathway_2021_Human\n",
      "2022-03-15 13:47:08,766 Save file of enrichment results: Job Id:9c6533edbca1941daa665648d9c2a4a9\n",
      "2022-03-15 13:47:08,909 Done.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WikiPathway_2021_Human shFOXA1_B14_vs_TGF_DOWN_genes\n",
      "i got res2d:                    Gene_set  \\\n",
      "0    WikiPathway_2021_Human   \n",
      "1    WikiPathway_2021_Human   \n",
      "2    WikiPathway_2021_Human   \n",
      "3    WikiPathway_2021_Human   \n",
      "4    WikiPathway_2021_Human   \n",
      "..                      ...   \n",
      "550  WikiPathway_2021_Human   \n",
      "551  WikiPathway_2021_Human   \n",
      "552  WikiPathway_2021_Human   \n",
      "553  WikiPathway_2021_Human   \n",
      "554  WikiPathway_2021_Human   \n",
      "\n",
      "                                                  Term  Overlap       P-value  \\\n",
      "0                VEGFA-VEGFR2 Signaling Pathway WP3888  123/432  6.757746e-25   \n",
      "1                                 Focal Adhesion WP306   61/198  1.053101e-14   \n",
      "2    Arrhythmogenic Right Ventricular Cardiomyopath...    30/74  3.035740e-11   \n",
      "3    Primary focal segmental glomerulosclerosis (FS...    28/72  4.361949e-10   \n",
      "4                   Ebola Virus Pathway on Host WP4217   39/129  1.193848e-09   \n",
      "..                                                 ...      ...           ...   \n",
      "550                              mRNA Processing WP411    5/126  9.983842e-01   \n",
      "551  DNA IR-damage and cellular response via ATR WP...     2/80  9.988391e-01   \n",
      "552                    Oxidative phosphorylation WP623     1/60  9.989320e-01   \n",
      "553               Cytoplasmic Ribosomal Proteins WP477     2/89  9.995410e-01   \n",
      "554  Electron Transport Chain (OXPHOS system in mit...    1/103  9.999889e-01   \n",
      "\n",
      "     Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
      "0        3.750549e-22            0                     0    3.438988   \n",
      "1        2.922355e-12            0                     0    3.769347   \n",
      "2        5.616118e-09            0                     0    5.717574   \n",
      "3        6.052204e-08            0                     0    5.331380   \n",
      "4        1.141811e-07            0                     0    3.639877   \n",
      "..                ...          ...                   ...         ...   \n",
      "550      9.999889e-01            0                     0    0.341005   \n",
      "551      9.999889e-01            0                     0    0.211816   \n",
      "552      9.999889e-01            0                     0    0.140098   \n",
      "553      9.999889e-01            0                     0    0.189807   \n",
      "554      9.999889e-01            0                     0    0.080841   \n",
      "\n",
      "     Combined Score                                              Genes  \n",
      "0      1.913932e+02  FHOD1;PLOD3;ETS1;CTGF;ACTG1;GJA1;CCND1;PLAU;CF...  \n",
      "1      1.213144e+02  ITGB1;ITGB5;ITGB4;PIK3CD;ILK;LAMC2;LAMC1;ACTB;...  \n",
      "2      1.384681e+02  ITGB1;ITGB5;ITGB4;TCF7;ACTB;ACTG1;GJA1;CACNG8;...  \n",
      "3      1.149069e+02  ITGB1;CDKN1A;NOTCH1;CD151;ITGB4;ILK;YWHAQ;CDH2...  \n",
      "4      7.478522e+01  ITGB1;CLTB;PIK3CD;IQGAP1;ACTB;RELA;EGFR;IGF1R;...  \n",
      "..              ...                                                ...  \n",
      "550    5.514352e-04                        PTBP1;HNRNPK;PRMT2;FUS;RBMX  \n",
      "551    2.460317e-04                                         H2AFX;TP53  \n",
      "552    1.497020e-04                                            ATP6AP2  \n",
      "553    8.714953e-05                                        RPL10;RPL28  \n",
      "554    9.002805e-07                                               UCP3  \n",
      "\n",
      "[555 rows x 10 columns]\n",
      "Legend is shown.localtion  best\n",
      "shNC_B14_vs_TGF_UP_genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-15 13:47:11,083 Connecting to Enrichr Server to get latest library names\n",
      "2022-03-15 13:47:11,084 Analysis name: test, Enrichr Library: WikiPathway_2021_Human\n",
      "2022-03-15 13:47:19,904 Save file of enrichment results: Job Id:72dae4a451fd09d4fa4b3f2680544307\n",
      "2022-03-15 13:47:20,054 Done.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WikiPathway_2021_Human shNC_B14_vs_TGF_UP_genes\n",
      "i got res2d:                    Gene_set  \\\n",
      "0    WikiPathway_2021_Human   \n",
      "1    WikiPathway_2021_Human   \n",
      "2    WikiPathway_2021_Human   \n",
      "3    WikiPathway_2021_Human   \n",
      "4    WikiPathway_2021_Human   \n",
      "..                      ...   \n",
      "553  WikiPathway_2021_Human   \n",
      "554  WikiPathway_2021_Human   \n",
      "555  WikiPathway_2021_Human   \n",
      "556  WikiPathway_2021_Human   \n",
      "557  WikiPathway_2021_Human   \n",
      "\n",
      "                                                  Term Overlap       P-value  \\\n",
      "0    Electron Transport Chain (OXPHOS system in mit...  62/103  7.213169e-32   \n",
      "1                 Cytoplasmic Ribosomal Proteins WP477   46/89  3.168857e-20   \n",
      "2              Nonalcoholic fatty liver disease WP4396  61/155  1.009858e-18   \n",
      "3       Metabolic reprogramming in colon cancer WP4290   29/42  4.938727e-18   \n",
      "4                      Oxidative phosphorylation WP623   35/60  5.822007e-18   \n",
      "..                                                 ...     ...           ...   \n",
      "553  Arrhythmogenic Right Ventricular Cardiomyopath...    2/74  9.989645e-01   \n",
      "554        Hippo-Merlin Signaling Dysregulation WP4541   5/120  9.989802e-01   \n",
      "555                                 Peptide GPCRs WP24    1/74  9.999021e-01   \n",
      "556                                 GPCRs, Other WP117    1/91  9.999855e-01   \n",
      "557                GPCRs, Class A Rhodopsin-like WP455   4/257  9.999953e-01   \n",
      "\n",
      "     Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
      "0        4.024948e-29            0                     0   11.643969   \n",
      "1        8.841112e-18            0                     0    8.179099   \n",
      "2        1.878335e-16            0                     0    4.979621   \n",
      "3        6.497360e-16            0                     0   16.959615   \n",
      "4        6.497360e-16            0                     0   10.663959   \n",
      "..                ...          ...                   ...         ...   \n",
      "553      9.999953e-01            0                     0    0.208055   \n",
      "554      9.999953e-01            0                     0    0.325271   \n",
      "555      9.999953e-01            0                     0    0.102553   \n",
      "556      9.999953e-01            0                     0    0.083101   \n",
      "557      9.999953e-01            0                     0    0.117299   \n",
      "\n",
      "     Combined Score                                              Genes  \n",
      "0      8.349519e+02  COX7B;COX4I1;NDUFA10;COX6A1;COX7C;UQCRFS1;COX8...  \n",
      "1      3.672279e+02  RPL4;RPL30;RPL3;RPL32;RPL31;RPL12;MRPL19;RPL10...  \n",
      "2      2.063392e+02  COX7B;COX4I1;NDUFA10;PRKAG2;IRS2;COX6A1;COX7C;...  \n",
      "3      6.758309e+02  GPI;FH;SHMT2;DLST;PDHB;SLC1A5;ENO1;PGK1;PSPH;I...  \n",
      "4      4.231980e+02  NDUFB9;NDUFB10;NDUFB6;NDUFB5;NDUFB4;NDUFA10;AT...  \n",
      "..              ...                                                ...  \n",
      "553    2.155445e-04                                       PKP2;CACNA1D  \n",
      "554    3.318851e-04                          DDB1;CDH1;TAZ;FGFR3;CDH17  \n",
      "555    1.003605e-05                                              NPY4R  \n",
      "556    1.206895e-06                                             CELSR2  \n",
      "557    5.487695e-07                           GPR37;OPN3;ADORA2B;NPY4R  \n",
      "\n",
      "[558 rows x 10 columns]\n",
      "Legend is shown.localtion  best\n",
      "shNC_B14_vs_TGF_DOWN_genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-15 13:47:22,294 Connecting to Enrichr Server to get latest library names\n",
      "2022-03-15 13:47:22,295 Analysis name: test, Enrichr Library: WikiPathway_2021_Human\n",
      "2022-03-15 13:47:31,526 Save file of enrichment results: Job Id:4f39ac1103c7d49f99f5030112d00dc7\n",
      "2022-03-15 13:47:31,672 Done.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WikiPathway_2021_Human shNC_B14_vs_TGF_DOWN_genes\n",
      "i got res2d:                    Gene_set  \\\n",
      "0    WikiPathway_2021_Human   \n",
      "1    WikiPathway_2021_Human   \n",
      "2    WikiPathway_2021_Human   \n",
      "3    WikiPathway_2021_Human   \n",
      "4    WikiPathway_2021_Human   \n",
      "..                      ...   \n",
      "553  WikiPathway_2021_Human   \n",
      "554  WikiPathway_2021_Human   \n",
      "555  WikiPathway_2021_Human   \n",
      "556  WikiPathway_2021_Human   \n",
      "557  WikiPathway_2021_Human   \n",
      "\n",
      "                                                  Term  Overlap       P-value  \\\n",
      "0                VEGFA-VEGFR2 Signaling Pathway WP3888  135/432  5.933577e-23   \n",
      "1                                 Focal Adhesion WP306   71/198  3.993827e-16   \n",
      "2                     EGF/EGFR signaling pathway WP437   52/162  3.756675e-10   \n",
      "3                Regulation of Actin Cytoskeleton WP51   49/150  6.094824e-10   \n",
      "4                     TGF-beta Signaling Pathway WP366   45/132  6.429959e-10   \n",
      "..                                                 ...      ...           ...   \n",
      "553               Cytoplasmic Ribosomal Proteins WP477     3/89  9.996522e-01   \n",
      "554                    Oxidative phosphorylation WP623     1/60  9.997965e-01   \n",
      "555                                 Peptide GPCRs WP24     1/74  9.999694e-01   \n",
      "556                GPCRs, Class A Rhodopsin-like WP455   13/257  9.999911e-01   \n",
      "557  Electron Transport Chain (OXPHOS system in mit...    1/103  9.999960e-01   \n",
      "\n",
      "     Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
      "0        3.310936e-20            0                     0    3.093339   \n",
      "1        1.114278e-13            0                     0    3.746822   \n",
      "2        6.987415e-08            0                     0    3.148108   \n",
      "3        7.175834e-08            0                     0    3.228773   \n",
      "4        7.175834e-08            0                     0    3.439844   \n",
      "..                ...          ...                   ...         ...   \n",
      "553      9.999960e-01            0                     0    0.228311   \n",
      "554      9.999960e-01            0                     0    0.111020   \n",
      "555      9.999960e-01            0                     0    0.089656   \n",
      "556      9.999960e-01            0                     0    0.346828   \n",
      "557      9.999960e-01            0                     0    0.064058   \n",
      "\n",
      "     Combined Score                                              Genes  \n",
      "0      1.583135e+02  FHOD1;PLOD3;ETS1;CTGF;ACTG1;GJA1;CCND1;PLAU;CF...  \n",
      "1      1.328496e+02  ITGB1;GSK3B;ITGB4;PIK3CD;ILK;LAMC2;LAMC1;ACTB;...  \n",
      "2      6.832124e+01  SH3KBP1;INPPL1;AP2A1;ELK4;GJA1;CFL1;GRB10;NCK2...  \n",
      "3      6.850943e+01  PIK3CD;FGF2;ARHGAP35;ACTB;ACTG1;TMSB4X;CFL1;PI...  \n",
      "4      7.280390e+01  ITGB1;YAP1;APP;CDKN1A;SHC1;ITGB4;TGFB1I1;NEDD9...  \n",
      "..              ...                                                ...  \n",
      "553    7.941119e-05                                  RPL10;RPLP1;RPL28  \n",
      "554    2.259552e-05                                            ATP6AP2  \n",
      "555    2.740652e-06                                              CXCR5  \n",
      "556    3.103835e-06  GPR39;F2R;HTR1D;CXCR5;LTB4R;HRH1;LPAR5;CNR1;GP...  \n",
      "557    2.572639e-07                                               UCP3  \n",
      "\n",
      "[558 rows x 10 columns]\n",
      "Legend is shown.localtion  best\n",
      "KEGG_2021_Human\n",
      "DMSO_nc_vs_kd_UP_genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-15 13:47:33,904 Connecting to Enrichr Server to get latest library names\n",
      "2022-03-15 13:47:33,905 Analysis name: test, Enrichr Library: KEGG_2021_Human\n",
      "2022-03-15 13:47:39,922 Save file of enrichment results: Job Id:6dc7c0cbb4d188f68b0c73b5fc6282c1\n",
      "2022-03-15 13:47:40,132 Done.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KEGG_2021_Human DMSO_nc_vs_kd_UP_genes\n",
      "i got res2d:             Gene_set                                     Term Overlap  \\\n",
      "0    KEGG_2021_Human                                 Lysosome  45/128   \n",
      "1    KEGG_2021_Human                    p53 signaling pathway   20/73   \n",
      "2    KEGG_2021_Human                              Ferroptosis   12/41   \n",
      "3    KEGG_2021_Human                     Steroid biosynthesis    8/20   \n",
      "4    KEGG_2021_Human                   Glutathione metabolism   12/57   \n",
      "..               ...                                      ...     ...   \n",
      "291  KEGG_2021_Human                            RNA transport   5/186   \n",
      "292  KEGG_2021_Human                mRNA surveillance pathway    1/98   \n",
      "293  KEGG_2021_Human                                 Ribosome   2/158   \n",
      "294  KEGG_2021_Human  Neuroactive ligand-receptor interaction   7/341   \n",
      "295  KEGG_2021_Human                   Olfactory transduction   1/440   \n",
      "\n",
      "          P-value  Adjusted P-value  Old P-value  Old Adjusted P-value  \\\n",
      "0    1.387045e-23      4.105652e-21            0                     0   \n",
      "1    4.564307e-09      6.755174e-07            0                     0   \n",
      "2    2.618913e-06      2.583995e-04            0                     0   \n",
      "3    9.361239e-06      6.927317e-04            0                     0   \n",
      "4    9.790624e-05      5.008243e-03            0                     0   \n",
      "..            ...               ...          ...                   ...   \n",
      "291  9.867427e-01      9.999955e-01            0                     0   \n",
      "292  9.974146e-01      9.999955e-01            0                     0   \n",
      "293  9.992706e-01      9.999955e-01            0                     0   \n",
      "294  9.998336e-01      9.999955e-01            0                     0   \n",
      "295  9.999955e-01      9.999955e-01            0                     0   \n",
      "\n",
      "     Odds Ratio  Combined Score  \\\n",
      "0      8.975478    4.723999e+02   \n",
      "1      6.121883    1.175708e+02   \n",
      "2      6.675388    8.579711e+01   \n",
      "3     10.727687    1.242152e+02   \n",
      "4      4.298255    3.967934e+01   \n",
      "..          ...             ...   \n",
      "291    0.439396    5.864188e-03   \n",
      "292    0.164159    4.249674e-04   \n",
      "293    0.203677    1.486214e-04   \n",
      "294    0.331191    5.512700e-05   \n",
      "295    0.035610    1.614469e-07   \n",
      "\n",
      "                                                 Genes  \n",
      "0    SCARB2;CD63;ASAH1;HEXB;CTSZ;HEXA;CLTA;LIPA;LIT...  \n",
      "1    CDKN1A;SERPINE1;TNFRSF10B;BBC3;CCND3;CASP8;CCN...  \n",
      "2    PRNP;SLC40A1;HMOX1;SAT2;SLC39A8;ACSL4;SLC7A11;...  \n",
      "3          SQLE;SC5D;DHCR24;MSMO1;DHCR7;LIPA;LSS;FDFT1  \n",
      "4    GSTK1;GSTM3;G6PD;GPX2;RRM2B;GPX3;MGST3;IDH1;GS...  \n",
      "..                                                 ...  \n",
      "291                 CYFIP2;EIF4A2;EIF2S3;PABPC4;EEF1A2  \n",
      "292                                             PABPC4  \n",
      "293                                       RPL22;RPS27L  \n",
      "294                C3;C5;GABBR1;GLRB;PARD3;S1PR3;TRPV1  \n",
      "295                                             CAMK2D  \n",
      "\n",
      "[296 rows x 10 columns]\n",
      "Legend is shown.localtion  best\n",
      "DMSO_nc_vs_kd_DOWN_genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-15 13:47:42,186 Connecting to Enrichr Server to get latest library names\n",
      "2022-03-15 13:47:42,187 Analysis name: test, Enrichr Library: KEGG_2021_Human\n",
      "2022-03-15 13:47:48,860 Save file of enrichment results: Job Id:fbf89756ef08e7b85346451c5258c7b2\n",
      "2022-03-15 13:47:48,998 Done.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KEGG_2021_Human DMSO_nc_vs_kd_DOWN_genes\n",
      "i got res2d:             Gene_set                                               Term  \\\n",
      "0    KEGG_2021_Human                                    DNA replication   \n",
      "1    KEGG_2021_Human                                        Spliceosome   \n",
      "2    KEGG_2021_Human                                         Cell cycle   \n",
      "3    KEGG_2021_Human                                    Mismatch repair   \n",
      "4    KEGG_2021_Human                               Base excision repair   \n",
      "..               ...                                                ...   \n",
      "276  KEGG_2021_Human                       Systemic lupus erythematosus   \n",
      "277  KEGG_2021_Human  Viral protein interaction with cytokine and cy...   \n",
      "278  KEGG_2021_Human                   Herpes simplex virus 1 infection   \n",
      "279  KEGG_2021_Human          Natural killer cell mediated cytotoxicity   \n",
      "280  KEGG_2021_Human                             Olfactory transduction   \n",
      "\n",
      "    Overlap       P-value  Adjusted P-value  Old P-value  \\\n",
      "0     24/36  8.333482e-21      2.341709e-18            0   \n",
      "1    46/150  5.511005e-20      7.742962e-18            0   \n",
      "2    40/124  1.963931e-18      1.839549e-16            0   \n",
      "3     16/23  9.740967e-15      6.843029e-13            0   \n",
      "4     14/33  4.049378e-09      2.275751e-07            0   \n",
      "..      ...           ...               ...          ...   \n",
      "276   2/135  9.985487e-01      9.999954e-01            0   \n",
      "277   1/100  9.985757e-01      9.999954e-01            0   \n",
      "278  15/498  9.997537e-01      9.999954e-01            0   \n",
      "279   1/131  9.998120e-01      9.999954e-01            0   \n",
      "280   2/440  9.999954e-01      9.999954e-01            0   \n",
      "\n",
      "     Old Adjusted P-value  Odds Ratio  Combined Score  \\\n",
      "0                       0   30.148148    1.393870e+03   \n",
      "1                       0    6.754256    2.995172e+02   \n",
      "2                       0    7.243844    2.953430e+02   \n",
      "3                       0   34.243657    1.104784e+03   \n",
      "4                       0   11.014377    2.128496e+02   \n",
      "..                    ...         ...             ...   \n",
      "276                     0    0.221293    3.214050e-04   \n",
      "277                     0    0.148800    2.120819e-04   \n",
      "278                     0    0.453079    1.115887e-04   \n",
      "279                     0    0.113129    2.127264e-05   \n",
      "280                     0    0.066094    3.061970e-07   \n",
      "\n",
      "                                                 Genes  \n",
      "0    RFC5;FEN1;RFC3;PCNA;RFC4;MCM7;LIG1;RFC1;RFC2;P...  \n",
      "1    SF3B4;TCERG1;SF3B3;EIF4A3;DDX23;SRSF1;DDX42;HN...  \n",
      "2    PCNA;MCM7;BUB1B;TTK;SMC3;PKMYT1;ANAPC11;CDC20;...  \n",
      "3    RFC5;RFC3;PCNA;RFC4;LIG1;RFC1;RFC2;RPA1;MLH1;M...  \n",
      "4    FEN1;PCNA;LIG1;PARP1;PARP2;XRCC1;HMGB1;UNG;POL...  \n",
      "..                                                 ...  \n",
      "276                                        ACTN4;SNRPB  \n",
      "277                                              IL6ST  \n",
      "278  SRC;EIF2AK1;ALYREF;SRSF1;TSC2;HCFC1;PPP1CA;OAS...  \n",
      "279                                               VAV2  \n",
      "280                                        PDE1C;ARRB2  \n",
      "\n",
      "[281 rows x 10 columns]\n",
      "Legend is shown.localtion  best\n",
      "TGF_nc_vs_kd_UP_genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-15 13:47:50,952 Connecting to Enrichr Server to get latest library names\n",
      "2022-03-15 13:47:50,953 Analysis name: test, Enrichr Library: KEGG_2021_Human\n",
      "2022-03-15 13:47:56,495 Save file of enrichment results: Job Id:2eb7e5aaebbcbbf933f612f4bc3600c8\n",
      "2022-03-15 13:47:56,632 Done.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KEGG_2021_Human TGF_nc_vs_kd_UP_genes\n",
      "i got res2d:             Gene_set                                     Term Overlap  \\\n",
      "0    KEGG_2021_Human                                 Lysosome  26/128   \n",
      "1    KEGG_2021_Human                    p53 signaling pathway   16/73   \n",
      "2    KEGG_2021_Human                 ECM-receptor interaction   17/88   \n",
      "3    KEGG_2021_Human                  Proteoglycans in cancer  26/205   \n",
      "4    KEGG_2021_Human                     Steroid biosynthesis    6/20   \n",
      "..               ...                                      ...     ...   \n",
      "280  KEGG_2021_Human  Neuroactive ligand-receptor interaction   7/341   \n",
      "281  KEGG_2021_Human                Oxidative phosphorylation   1/133   \n",
      "282  KEGG_2021_Human                            RNA transport   2/186   \n",
      "283  KEGG_2021_Human                                 Ribosome   1/158   \n",
      "284  KEGG_2021_Human                               Alcoholism   1/186   \n",
      "\n",
      "          P-value  Adjusted P-value  Old P-value  Old Adjusted P-value  \\\n",
      "0    3.271102e-11      9.322640e-09            0                     0   \n",
      "1    6.485585e-08      9.241959e-06            0                     0   \n",
      "2    1.799610e-07      1.709629e-05            0                     0   \n",
      "3    8.458740e-07      6.026852e-05            0                     0   \n",
      "4    1.451540e-04      7.935637e-03            0                     0   \n",
      "..            ...               ...          ...                   ...   \n",
      "280  9.922064e-01      9.997314e-01            0                     0   \n",
      "281  9.972008e-01      9.997314e-01            0                     0   \n",
      "282  9.974845e-01      9.997314e-01            0                     0   \n",
      "283  9.990752e-01      9.997314e-01            0                     0   \n",
      "284  9.997314e-01      9.997314e-01            0                     0   \n",
      "\n",
      "     Odds Ratio  Combined Score  \\\n",
      "0      5.804203      140.132668   \n",
      "1      6.331052      104.785870   \n",
      "2      5.402767       83.907811   \n",
      "3      3.294045       46.060280   \n",
      "4      9.574766       84.619056   \n",
      "..          ...             ...   \n",
      "280    0.460931        0.003606   \n",
      "281    0.167230        0.000469   \n",
      "282    0.239560        0.000603   \n",
      "283    0.140416        0.000130   \n",
      "284    0.118988        0.000032   \n",
      "\n",
      "                                                 Genes  \n",
      "0    CD63;ASAH1;HEXA;CLTA;TCIRG1;AP1G2;GM2A;CTSL;HY...  \n",
      "1    CDKN1A;SERPINE1;TNFRSF10B;BBC3;CCND3;TP53I3;SE...  \n",
      "2    LAMA5;ITGB5;ITGA4;ITGB4;ITGA3;LAMA4;LAMA3;FN1;...  \n",
      "3    CD63;CDKN1A;ITGB5;PXN;ITPR3;HIF1A;FGF2;VTN;CTS...  \n",
      "4                    SQLE;DHCR24;MSMO1;DHCR7;LSS;FDFT1  \n",
      "..                                                 ...  \n",
      "280               C3;PARD3;HTR1D;TSPO;ADM;TRPV1;ADRA1B  \n",
      "281                                             TCIRG1  \n",
      "282                                      CYFIP2;EEF1A2  \n",
      "283                                             RPS27L  \n",
      "284                                            SLC29A2  \n",
      "\n",
      "[285 rows x 10 columns]\n",
      "Legend is shown.localtion  best\n",
      "TGF_nc_vs_kd_DOWN_genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-15 13:47:58,648 Connecting to Enrichr Server to get latest library names\n",
      "2022-03-15 13:47:58,649 Analysis name: test, Enrichr Library: KEGG_2021_Human\n",
      "2022-03-15 13:48:04,124 Save file of enrichment results: Job Id:9fb07eb23d04a557897781e735756eb0\n",
      "2022-03-15 13:48:04,258 Done.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KEGG_2021_Human TGF_nc_vs_kd_DOWN_genes\n",
      "i got res2d:             Gene_set                                     Term Overlap  \\\n",
      "0    KEGG_2021_Human                               Cell cycle  39/124   \n",
      "1    KEGG_2021_Human                          DNA replication   21/36   \n",
      "2    KEGG_2021_Human                          Mismatch repair   11/23   \n",
      "3    KEGG_2021_Human                   Fanconi anemia pathway   14/54   \n",
      "4    KEGG_2021_Human                     Base excision repair   11/33   \n",
      "..               ...                                      ...     ...   \n",
      "244  KEGG_2021_Human                        Alzheimer disease   5/369   \n",
      "245  KEGG_2021_Human  Neuroactive ligand-receptor interaction   4/341   \n",
      "246  KEGG_2021_Human  Neutrophil extracellular trap formation   1/189   \n",
      "247  KEGG_2021_Human         Herpes simplex virus 1 infection   3/498   \n",
      "248  KEGG_2021_Human                   Olfactory transduction   2/440   \n",
      "\n",
      "          P-value  Adjusted P-value  Old P-value  Old Adjusted P-value  \\\n",
      "0    2.479573e-25      6.174136e-23            0                     0   \n",
      "1    4.256254e-21      5.299036e-19            0                     0   \n",
      "2    2.120772e-10      1.760241e-08            0                     0   \n",
      "3    9.985827e-09      6.216178e-07            0                     0   \n",
      "4    2.138054e-08      1.064751e-06            0                     0   \n",
      "..            ...               ...          ...                   ...   \n",
      "244  9.986641e-01      9.999955e-01            0                     0   \n",
      "245  9.991871e-01      9.999955e-01            0                     0   \n",
      "246  9.993899e-01      9.999955e-01            0                     0   \n",
      "247  9.999952e-01      9.999955e-01            0                     0   \n",
      "248  9.999955e-01      9.999955e-01            0                     0   \n",
      "\n",
      "     Odds Ratio  Combined Score  \\\n",
      "0     12.102577    6.856901e+02   \n",
      "1     36.166667    1.696430e+03   \n",
      "2     23.370137    5.205481e+02   \n",
      "3      8.945739    1.647993e+02   \n",
      "4     12.740716    2.250110e+02   \n",
      "..          ...             ...   \n",
      "244    0.341075    4.559309e-04   \n",
      "245    0.294755    2.396950e-04   \n",
      "246    0.132610    8.092629e-05   \n",
      "247    0.149050    7.120900e-07   \n",
      "248    0.112492    5.075009e-07   \n",
      "\n",
      "                                                 Genes  \n",
      "0    PCNA;MCM7;BUB1B;TTK;SMC3;PKMYT1;CDC20;CCNB2;CC...  \n",
      "1    RFC5;PRIM2;FEN1;RFC3;PCNA;RFC4;MCM7;LIG1;PRIM1...  \n",
      "2    RFC5;MSH6;POLD3;RFC3;PCNA;RFC4;LIG1;MSH2;EXO1;...  \n",
      "3    FANCI;RMI2;BLM;RMI1;FANCA;BRCA1;FANCB;BRCA2;FA...  \n",
      "4    POLD3;FEN1;NEIL3;PCNA;LIG1;PARP1;POLD1;TDG;POL...  \n",
      "..                                                 ...  \n",
      "244                    TUBA1B;PSMD2;KIF5B;WNT5A;ATP2A3  \n",
      "245                         GABRB3;GABRQ;CHRNA5;ADRA1D  \n",
      "246                                              HMGB1  \n",
      "247                                   OAS3;BCL2;ZNF853  \n",
      "248                                         RGS2;ARRB2  \n",
      "\n",
      "[249 rows x 10 columns]\n",
      "Legend is shown.localtion  best\n",
      "shFOXA1_B14_vs_TGF_UP_genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-15 13:48:06,228 Connecting to Enrichr Server to get latest library names\n",
      "2022-03-15 13:48:06,230 Analysis name: test, Enrichr Library: KEGG_2021_Human\n"
     ]
    }
   ],
   "source": [
    "for lib_name in target_libs:\n",
    "    print(lib_name)\n",
    "    \n",
    "    for gene_list_name, gene_list in dict_gene_list.items():\n",
    "        print(gene_list_name)\n",
    "        # print(gene_list)\n",
    "        try :\n",
    "            enr = gp.enrichr(\n",
    "                        gene_list=gene_list,\n",
    "                        gene_sets=lib_name,\n",
    "                        organism=speices,\n",
    "                        description=description,\n",
    "                        outdir=wk_dir+'barplots_and_tabs/'+gene_list_name,\n",
    "                        # no_plot=False,\n",
    "                        cutoff=0.05,\n",
    "                        verbose=True\n",
    "                        )\n",
    "            # print(enr.res2d)   \n",
    "            # time.sleep(10)\n",
    "        except Exception as e:\n",
    "            print('error:', e)\n",
    "\n",
    "        ## to save your figure, make sure that ``ofname`` is not None\n",
    "        # title = ' and '.join(target_libs) + ' ' + gene_list_name\n",
    "        title = lib_name + ' ' + gene_list_name\n",
    "        print(title)\n",
    "\n",
    "        try :\n",
    "            print('i got res2d: {0}'.format(enr.res2d))\n",
    "            dotplot(enr.res2d,title=title,\n",
    "                top_term=20,\n",
    "                cmap='viridis_r',\n",
    "                ofname=wk_dir + title+'.png',\n",
    "                )\n",
    "            # time.sleep(10)                \n",
    "        except Exception as e2 :\n",
    "            print('error2: ',e2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Also could input target_libs whole list to gene_sets; eg gene_sets=target_libs, this will combine resluts to one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene_list_name, gene_list in dict_gene_list.items():\n",
    "    print(gene_list_name)\n",
    "\n",
    "name = 'kd_B14_vs_TGFb_UP_genes'\n",
    "list = dict_gene_list[name]\n",
    "enr = gp.enrichr(\n",
    "                gene_list=list,\n",
    "                gene_sets=target_libs,\n",
    "                organism='Human',\n",
    "                description=description,\n",
    "                outdir=wk_dir+'enrichr_plots/'+name,\n",
    "                # no_plot=False,\n",
    "                cutoff=0.05,\n",
    "                verbose=True\n",
    "                )\n",
    "# print(enr.res2d)   \n",
    "\n",
    "\n",
    "# to save your figure, make sure that ``ofname`` is not None\n",
    "title = ' and '.join(target_libs) + ' ' + name\n",
    "# print(title)\n",
    "\n",
    "dotplot(enr.res2d,title=title,\n",
    "        top_term=20,\n",
    "        cmap='viridis_r',\n",
    "        # ofname=wk_dir + title+'.png',\n",
    "        )   "
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "40d3a090f54c6569ab1632332b64b2c03c39dcf918b08424e98f38b5ae0af88f"
  },
  "kernelspec": {
   "display_name": "Python 3.8.5 64-bit ('base': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
